SpatialGDKEditorToolbar/Source/Private/SchemaGenerator/SchemaGenerator.cpp

```diff
@@ -298,5 +438,185 @@ int GenerateTypeBindingSchema(FCodeWriter& Writer, int ComponentId, UClass* Clas
+{
+	FUnrealFlatRepData RepData = GetFlatRepData(TypeInfo);
+
+	FSubobjectSchemaData SubobjectData;
+	SubobjectData.Class = ComponentClass;
+
+	for (EReplicatedPropertyGroup Group : GetAllReplicatedPropertyGroups())
+	{
+		if (RepData[Group].Num() == 0)
+		{
+			continue;
+		}
+
+		Writer.PrintNewLine();
+
+		FString ComponentName = PropertyName + GetReplicatedPropertyGroupName(Group);
+		Writer.Printf("component {0} {", *ComponentName);
+		Writer.Indent();
+		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		Writer.Printf("data {0};", *SchemaReplicatedDataName(Group, ComponentClass));
+		Writer.Outdent().Print("}");
+
+		SubobjectData.SchemaComponents[PropertyGroupToSchemaComponentType(Group)] = IdGenerator.GetCurrentId();
+	}
+
+	FCmdHandlePropertyMap HandoverData = GetFlatHandoverData(TypeInfo);
+	if (HandoverData.Num() > 0)
+	{
+		Writer.PrintNewLine();
+
+		// Handover (server to server) replicated properties.
+		Writer.Printf("component {0} {", *(PropertyName + TEXT("Handover")));
+		Writer.Indent();
+		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		Writer.Printf("data {0};", *SchemaHandoverDataName(ComponentClass));
+		Writer.Outdent().Print("}");
+
+		SubobjectData.SchemaComponents[ESchemaComponentType::SCHEMA_Handover] = IdGenerator.GetCurrentId();
+	}
+
+	FUnrealRPCsByType RPCsByType = GetAllRPCsByType(TypeInfo);
+
+	for (auto Group : GetRPCTypes())
+	{
+		if (RPCsByType[Group].Num() == 0)
+		{
+			continue;
+		}
+
+		Writer.PrintNewLine();
+
+		FString ComponentName = PropertyName + GetRPCTypeName(Group) + TEXT("RPCs");
+		Writer.Printf("component {0} {", *ComponentName);
+		Writer.Indent();
+		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		for (auto& RPC : RPCsByType[Group])
+		{
+			if (Group == ERPCType::RPC_NetMulticast)
+			{
+				Writer.Printf("event UnrealRPCCommandRequest {0};",
+					*SchemaRPCName(Cast<UClass>(ComponentClass), RPC->Function));
+			}
+			else
+			{
+				Writer.Printf("command UnrealRPCCommandResponse {0}(UnrealRPCCommandRequest);",
+					*SchemaRPCName(Cast<UClass>(ComponentClass), RPC->Function));
+			}
+		}
+		Writer.Outdent().Print("}");
+
+		SubobjectData.SchemaComponents[RPCTypeToSchemaComponentType(Group)] = IdGenerator.GetCurrentId();
+	}
+
+	return SubobjectData;
+}
+
+void GenerateSubobjectSchemaForActor(FComponentIdGenerator& IdGenerator, UClass* ActorClass, TSharedPtr<FUnrealType> TypeInfo, FString SchemaPath, FSchemaData& ActorSchemaData)
+{
+	FCodeWriter Writer;
+
+	Writer.Printf(R"""(
+		// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+		// Note that this file has been generated automatically
+		package unreal.generated.{0}.components;)""",
+		*UnrealNameToSchemaTypeName(TypeInfo->Type->GetName().ToLower()));
+
+	Writer.PrintNewLine();
+
+	GenerateActorIncludes(Writer, TypeInfo);
+
+	bool bHasComponents = false;
+	TSet<UObject*> SeenComponents;
+	int32 CurrentOffset = 1;
+
+	for (auto& PropertyPair : TypeInfo->Properties)
```

Comment:
Iterating the properties seems a bit weird at this point, considering all the subobjects we need should live on the CDO or have been created by us (the blueprint components).
So maybe `TypeInfo` should already have a `StaticSubobjects` array containing all of them at this point.